 # TP TUNING
 
 ## Prerequisites
 In this section, we will cover different aspects about the optimizer.
 
 Before going on, please connect you to the database.
 

In [29]:
-- connection: host='localhost' dbname='ds2' user='ds2'

## <span style="color:blue">EX - 1</span>

In this exercise, we will see what happens when you don't use index or any other optimization such as partitionning.


All statistics are stored in postgres into a view catalog called pg_stats. Another view is available trough pg_statistic but this one is less easy to read.<br/>
Click on <a href="https://www.postgresql.org/docs/11/view-pg-stats.html">pg_stats</a> to get a complete description.

In our first step, we will focus on a test table we will create.


In [36]:
SET max_parallel_workers_per_gather TO 0;
DROP TABLE IF EXISTS test;
CREATE TABLE test (i integer not null, t text);
explain (ANALYZE , TIMING ON )INSERT INTO test SELECT CASE WHEN i > 700000 THEN 700000 ELSE i/1000 END, md5(i::text) FROM generate_series(1, 1000000) i;

4 row(s) returned.


QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------
Insert on test  (cost=0.00..22.50 rows=1000 width=36) (actual time=1417.053..1417.053 rows=0 loops=1)
->  Function Scan on generate_series i  (cost=0.00..22.50 rows=1000 width=36) (actual time=74.285..511.145 rows=1000000 loops=1)
Planning Time: 0.039 ms
Execution Time: 1418.554 ms

Note that our data are inserted in 1365.778 ms : depends on your computer.<br/>
In order to get the description of the table, run the following command :

In [37]:
SELECT *
FROM information_schema.columns
WHERE table_schema = 'ds2'
  AND table_name   = 'test';

2 row(s) returned.


table_catalog    table_schema    table_name    column_name      ordinal_position  column_default    is_nullable    data_type    character_maximum_length      character_octet_length    numeric_precision    numeric_precision_radix    numeric_scale  datetime_precision    interval_type    interval_precision    character_set_catalog    character_set_schema    character_set_name    collation_catalog    collation_schema    collation_name    domain_catalog    domain_schema    domain_name    udt_catalog    udt_schema    udt_name    scope_catalog    scope_schema    scope_name    maximum_cardinality      dtd_identifier  is_self_referencing    is_identity    identity_generation    identity_start    identity_increment    identity_maximum    identity_minimum    identity_cycle    is_generated    generation_expression    is_updatable
---------------  --------------  ------------  -------------  ------------------  ----------------  -------------  -----------  --------------------------  ------------------------  -------------------  -------------------------  ---------------  --------------------  ---------------  --------------------  -----------------------  ----------------------  --------------------  -------------------  ------------------  ----------------  ----------------  ---------------  -------------  -------------  ------------  ----------  ---------------  --------------  ------------  ---------------------  ----------------  ---------------------  -------------  ---------------------  ----------------  --------------------  ------------------  ------------------  ----------------  --------------  -----------------------  --------------
ds2              ds2             test          i                               1                    NO             integer                                                                             32                          2                0                                                                                                                                                                                                                                                   ds2            pg_catalog    int4                                                                                             1  NO                     NO                                                                                                                    NO                NEVER                                    YES
ds2              ds2             test          t                               2                    YES            text                                                   1073741824                                                                                                                                                                                                                                                                                                                    ds2            pg_catalog    text                                                                                             2  NO                     NO                                                                                                                    NO                NEVER                                    YES

To get the index information on the test table :

In [38]:
SELECT * from pg_indexes where  tablename='test';

0 row(s) returned.


To update statistics on our new test table, use the ANALYZE command :

In [39]:
ANALYZE test;

Let's have a look on statistics of the test table : 

In [40]:
select * from pg_stats where schemaname ='ds2' and tablename='test';


2 row(s) returned.


schemaname    tablename    attname    inherited      null_frac    avg_width    n_distinct  most_common_vals                            most_common_freqs                                                                                                   histogram_bounds                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          correlation  most_common_elems    most_common_elem_freqs    elem_count_histogram
------------  -----------  ---------  -----------  -----------  -----------  ------------  ------------------------------------------  ------------------------------------------------------------------------------------------------------------------  -------------------------------------------------------------------------------

You can get a full field description here :
https://www.postgresql.org/docs/11/view-pg-stats.html<br/>

Great, our statistics are up-to-date.<br/>

Let's look at how the RDBMS estimates the cost of a SELECT.<br/>

Postgres defines a cost for each operation :
* cpu_tuple_cost : cost of processing each row during a query. The default is 0.01.
* cpu_operator_cost : cost of processing each operator or function executed during a query. The default is 0.0025.
* seq_page_cost :  cost of a disk page fetch that is part of a series of sequential fetches. The default is 1.0.<br/>

A complete description of postgresql variables is available here :https://www.postgresql.org/docs/11/runtime-config-query.html

To estimate the cost, the optimizer will take into account : 
* system variables
* number of pages to analyze (relpages)
* number of row to analyze (reltuples)

For example, when you run a full scan of your table via the following SQL query :

In [43]:
explain analyze select * from test;

3 row(s) returned.


QUERY PLAN
---------------------------------------------------------------------------------------------------------------
Seq Scan on test  (cost=0.00..18334.00 rows=1000000 width=37) (actual time=0.009..167.979 rows=1000000 loops=1)
Planning Time: 0.027 ms
Execution Time: 259.614 ms

Even though the optimizer knows how data is distributed, it has only one way to scan data: sequentially.<br/>
Of course, this method will be used whatever the value of your predicat is.<br/>

We agree the RDBMS will need to read sequentially all table pages and consequently all rows.<br/>
So the cost will be based on the :
* cost of processing each row : the total number of rows * cpu_tuple_cost(0,01)
* cost of a disk page fetch :  the total number of pages *  seq_page_cost(1)
* cost of a query processing :  the total number of pages *  cpu_operator_cost(0,0025)

If we addition all these costs we get the same estimate than the optimizer.

An estimate of the number of pages and rows are stored in the pg_class view into reltuples and relpages columns.

In [44]:
select to_char(reltuples,'999 999 999 999') reltuples, relpages from pg_class where relname ='test';

1 row(s) returned.


reltuples      relpages
-----------  ----------
1 000 000          8334

To get the estimate cost of you "select * from test; ", we can find it by runing the following query :

In [47]:
select relname,relpages * current_setting('seq_page_cost')::float + 
reltuples * current_setting('cpu_tuple_cost')::float + 
current_setting('cpu_operator_cost')::float "Estimated cost"
from pg_class where relname = 'test';

1 row(s) returned.


relname      Estimated cost
---------  ----------------
test                  18334

So, according to our compute, reading the whole data costs 18334 and it's the estimate we got in the explain command.

If you look at the estimate cost with a predicat you could notice this one is higher than our previous query :

In [48]:
explain analyze select * from test where i =56;

5 row(s) returned.


QUERY PLAN
-------------------------------------------------------------------------------------------------------
Seq Scan on test  (cost=0.00..20834.00 rows=995 width=37) (actual time=5.385..75.641 rows=1000 loops=1)
Filter: (i = 56)
Rows Removed by Filter: 999000
Planning Time: 0.351 ms
Execution Time: 75.738 ms

Could you explain why ?

## <span style="color:blue">EX - 2</span>

In this exercise, we will see how postgresql use a single index.

In our first step, we will focus on a test table we will create with an index.

In [71]:
DROP TABLE IF EXISTS test_ex2;
CREATE TABLE test_ex2 (i integer not null, t text);
INSERT INTO test_ex2 SELECT i/1000 , md5(i::text) FROM generate_series(1, 1000000) i;
CREATE INDEX ON test_ex2 (i);

In order to get the description of the table, run the following command :

In [72]:
SELECT *
FROM information_schema.columns
WHERE table_schema = 'ds2'
  AND table_name   = 'test_ex2';

2 row(s) returned.


table_catalog    table_schema    table_name    column_name      ordinal_position  column_default    is_nullable    data_type    character_maximum_length      character_octet_length    numeric_precision    numeric_precision_radix    numeric_scale  datetime_precision    interval_type    interval_precision    character_set_catalog    character_set_schema    character_set_name    collation_catalog    collation_schema    collation_name    domain_catalog    domain_schema    domain_name    udt_catalog    udt_schema    udt_name    scope_catalog    scope_schema    scope_name    maximum_cardinality      dtd_identifier  is_self_referencing    is_identity    identity_generation    identity_start    identity_increment    identity_maximum    identity_minimum    identity_cycle    is_generated    generation_expression    is_updatable
---------------  --------------  ------------  -------------  ------------------  ----------------  -------------  -----------  --------------------------  ------------------------  -------------------  -------------------------  ---------------  --------------------  ---------------  --------------------  -----------------------  ----------------------  --------------------  -------------------  ------------------  ----------------  ----------------  ---------------  -------------  -------------  ------------  ----------  ---------------  --------------  ------------  ---------------------  ----------------  ---------------------  -------------  ---------------------  ----------------  --------------------  ------------------  ------------------  ----------------  --------------  -----------------------  --------------
ds2              ds2             test_ex2      i                               1                    NO             integer                                                                             32                          2                0                                                                                                                                                                                                                                                   ds2            pg_catalog    int4                                                                                             1  NO                     NO                                                                                                                    NO                NEVER                                    YES
ds2              ds2             test_ex2      t                               2                    YES            text                                                   1073741824                                                                                                                                                                                                                                                                                                                    ds2            pg_catalog    text                                                                                             2  NO                     NO                                                                                                                    NO                NEVER                                    YES

To get the index information on the test_ex2 table :

In [73]:
SELECT * from pg_indexes where  tablename='test_ex2';

1 row(s) returned.


schemaname    tablename    indexname       tablespace    indexdef
------------  -----------  --------------  ------------  -----------------------------------------------------------
ds2           test_ex2     test_ex2_i_idx                CREATE INDEX test_ex2_i_idx ON ds2.test_ex2 USING btree (i)

To update statistics on our new test table, use the ANALYZE command :

In [74]:
ANALYZE test_ex2;

For each object of the ds2 schema, we can see statistics which are collected with the auto-vacuum or the ANALYZE command.
In this practice, we will limit statistics to the test table :

In [75]:
select * from pg_stats where schemaname ='ds2' and tablename='test_ex2';

2 row(s) returned.


schemaname    tablename    attname    inherited      null_frac    avg_width    n_distinct  most_common_vals                              most_common_freqs                                                                                                             histogram_bounds                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          correlation  most_common_elems    most_common_elem_freqs    elem_count_histogram
------------  -----------  ---------  -----------  -----------  -----------  ------------  --------------------------------------------  ----------------------------------------------------------------------------------------------------------------------------  -------------------------------------------------------

Now, let's compute the selectivity on the i column.<br/>
The RDBMS looks at the data distribution through most_common_vals and most_common_freqs column.<br/>
Most_common_freqs give the selectivity and most_common_vals give the value associated to the selectivity.

In [84]:
SELECT tablename, attname, value, freq selectivity, freq * 1000000 cardinality 
FROM (SELECT tablename, attname, mcv.value, mcv.freq FROM pg_stats,
LATERAL ROWS FROM (unnest(most_common_vals::text::int[]), unnest(most_common_freqs)) AS mcv(value, freq)
WHERE tablename = 'test_ex2'
AND attname = 'i') get_mcv order by freq DESC limit 5 ;


5 row(s) returned.


tablename    attname      value    selectivity    cardinality
-----------  ---------  -------  -------------  -------------
test_ex2     i              307     0.00103556        1035.56
test_ex2     i                3     0.00103111        1031.11
test_ex2     i              164     0.00103           1030
test_ex2     i              816     0.00103           1030
test_ex2     i              524     0.00102889        1028.89

According to the result, we identified the selectivity for each i value but is it relevant ? <br/>
Of course, that depends on the predicat we use in our query.

Let's take an example with the predicat "i = x" where x is a value of the result from the previous query.<br/>
<span style="color:red">
Watch out, statistics are based on a sample so the selectivity can differ if you rerun an ANALYZE command.</span>

So, the optimizer should estimate the number of rows as follow : <br/>
Selectivity * 1 000 000

Now let us check what the optimizer will estimate : 
Replace the x variable  with your value.

In [78]:
explain (ANALYZE, BUFFERS, TIMING ON, VERBOSE) select * from  test_ex2 where i = x; -- CHANGE X

6 row(s) returned.


QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------
Index Scan using test_ex2_i_idx on ds2.test_ex2  (cost=0.42..62.67 rows=1500 width=37) (actual time=0.033..0.206 rows=1000 loops=1)
Output: i, t
Index Cond: (test_ex2.i = 329)
Buffers: shared hit=10 read=5
Planning Time: 0.110 ms
Execution Time: 0.324 ms


From the first row, our computed cardinality matches with the the optimtizer estimate.<br/>

You may compute the estimate error with the following formula :

\| actual row - row estimate |<br/>
 \---------------------------<br/>
            actual row <br/>

According to the sampling, the error may be more or less important.<br/>
Don't forget, sampling is random and may not represent correctly your data.<br/>
<br/>
Postgres by default, takes a sample of 300 * default_statistics_target where default_statistics_target =100.</br>
This parameter is defined for any table, so you can customize statistics per table.<br/>

If we want to impove the selectivity score we should increase the sample size by increasing the default_statistics_target to 10000, in order to cover the whole dataset.


In [95]:
ALTER table test_ex2 ALTER COLUMN i SET STATISTICS 10000;

In [96]:
ANALYZE test_ex2;

Now, can you recompute the  new selectivity and cardinality :


In [98]:
SELECT tablename, attname, value, freq selectivity, freq * 1000000 cardinality 
FROM (SELECT tablename, attname, mcv.value, mcv.freq FROM pg_stats,
LATERAL ROWS FROM (unnest(most_common_vals::text::int[]), unnest(most_common_freqs)) AS mcv(value, freq)
WHERE tablename = 'test_ex2'
AND attname = 'i') get_mcv order by freq DESC limit 5 ;

5 row(s) returned.


tablename    attname      value    selectivity    cardinality
-----------  ---------  -------  -------------  -------------
test_ex2     i                2          0.001           1000
test_ex2     i                3          0.001           1000
test_ex2     i                4          0.001           1000
test_ex2     i                5          0.001           1000
test_ex2     i                1          0.001           1000

In [99]:
explain (ANALYZE, BUFFERS, TIMING ON, VERBOSE) select * from  test_ex2 where i = x ; -- CHANGE X

6 row(s) returned.


QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------
Index Scan using test_ex2_i_idx on ds2.test_ex2  (cost=0.42..41.92 rows=1000 width=37) (actual time=0.010..0.390 rows=1000 loops=1)
Output: i, t
Index Cond: (test_ex2.i = 307)
Buffers: shared hit=15
Planning Time: 0.114 ms
Execution Time: 0.488 ms

Can you compute the new error ? What do you think ?

Perfect, now our estimation is exactly the number of actual rows.<br/>
The estimate error is now to 0 but at what cost!
Indeed, the larger the sample is, the more the ANALYZE process will take time but in return we improve the planner's estimates quality.<br/>
Finally, the right question is what is the default_statistics_target value for representing data ? 



From statistics, the optimizer is able to estimate how many rows will be returned with a predicat "=" and consequently may choose an adapted access path.
<br/>In our use case, the selectivity is strong (close to 0) so the best solution to scan and collect the result is to browse the index.<br/>

## <span style="color:blue">EX - 3</span>

In this exercice, we will see how histogram is used.<br/>
Histogram is used to determine the selectivity when the number of distinct value is greater then 250.</br>
Indeed, listing all frequent value would be too large to be kept into most_common_freqs so histogram is a good alternative to describe data distribution.<br/>
Firstly, you will create a new test table with an index.

In [196]:
DROP TABLE IF EXISTS testhistogram;
CREATE TABLE testhistogram (i integer not null, t text);
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
INSERT INTO testhistogram SELECT i, md5(i::text) FROM generate_series(1, 1000) i;
CREATE INDEX ON testhistogram (i);

The number of buckets is set up from the default_statistics_target variable.<br/>
To customize the number of bucket, run the following command :

In [197]:
ALTER table testhistogram ALTER COLUMN i SET STATISTICS 10;

We inserted 10 000 rows with 1 000 distinct values into the test table, so the bucket width should be equal to 100.<br/>
Each bucket should contain 100 rows per bucket.

In [207]:
analyze testhistogram;
select histogram_bounds,most_common_vals, most_common_freqs from pg_stats where schemaname ='ds2' and tablename='testhistogram' and attname='i' ;

1 row(s) returned.


histogram_bounds                               most_common_vals        most_common_freqs
---------------------------------------------  ----------------------  ----------------------------------------------------------------------
{11,109,208,307,406,505,604,703,802,901,1000}  {1,2,3,4,5,6,7,8,9,10}  [0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001]

In my demonstration I got the following result :<br/>
{11,109,208,307,406,505,604,703,802,901,1000}.<br/>
So, the first bucket contains 100 values between 11 and 109, 100  values between 109 and 208 and so on.<br/>

Could you identify your bucket structure ?


From histogram, we want to determine the selectivity of a query where the predicat is "i < 300".<br/>
To do that, Postgres details how to compute them on the official documentation : <br/>
https://www.postgresql.org/docs/11/row-estimation-examples.html<br/>

In my case, the compute is the following :<br/>
selectivity = (1 + (300 - bucket[2].min)/(bucket[2].max - bucket[2].min))/num_buckets<br/>
selectivity = (1 + (300 - 208)/(307 - 208))::float/10<br/>
selectivity = 0.1<br/>



In [215]:
select (1 + (300 - 208)/(307 - 208))::float/10 Selectivity, ((1 + (300 - 208)/(307 - 208))::float/10)* 10000 Cardinality;

1 row(s) returned.


selectivity    cardinality
-------------  -------------
          0.1           1000

From my sample, I get a cardinality of 1000 with my predicat < 300.<br/>
Now, let's check if we get the same cardinality as the optimizer :

In [216]:
EXPLAIN (analyze)SELECT * FROM testhistogram WHERE  i < 100;

7 row(s) returned.


QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------
Bitmap Heap Scan on testhistogram  (cost=20.02..116.49 rows=998 width=37) (actual time=0.044..0.270 rows=990 loops=1)
Recheck Cond: (i < 100)
Heap Blocks: exact=16
->  Bitmap Index Scan on testhistogram_i_idx  (cost=0.00..19.77 rows=998 width=0) (actual time=0.036..0.036 rows=990 loops=1)
Index Cond: (i < 100)
Planning Time: 0.076 ms
Execution Time: 0.396 ms

The optimizer returns a cardinality of 998 which is close to my selectivity estimate (1000) but is not identical.<br/>
The optimizer should have taken some additional informations to adjust its cardinality but it is not possible to know which one.<br/>
Indeed, its cardinality (998) is closer to the actual value (990).

------------------------------------------------------------------------------------------------------------------------------

## <span style="color:blue">EX - 4</span>

In this exercice, we will see how to influence the optimizer.

Create a new test table as indicated below :


In [223]:
DROP TABLE IF EXISTS test_ex2;
CREATE TABLE test_ex2 (i integer not null, t text);
INSERT INTO test_ex2 SELECT i , md5(i::text) FROM generate_series(1, 1000000) i;
CREATE INDEX ON test_ex2 (i);
ANALYZE test_ex2;

In [224]:
SET enable_bitmapscan=on;
SET enable_indexscan=on; 
SET enable_seqscan=on;

Now run a query with a predicat "i = 100"

In [226]:
EXPLAIN (analyze)SELECT * FROM test_ex2 WHERE  i = 100;

4 row(s) returned.


QUERY PLAN
------------------------------------------------------------------------------------------------------------------------
Index Scan using test_ex2_i_idx on test_ex2  (cost=0.42..8.44 rows=1 width=37) (actual time=0.021..0.022 rows=1 loops=1)
Index Cond: (i = 100)
Planning Time: 0.224 ms
Execution Time: 0.042 ms

What is the cost of the query ? <br/>
Do you agree with the optimizer's choice ? <br/>

We are going to force the optimizer to avoid using the index scan method :


In [227]:
SET enable_indexscan=off; 

In [ ]:
Rerun the query :

In [228]:
EXPLAIN (analyze)SELECT * FROM test_ex2 WHERE  i = 100;

7 row(s) returned.


QUERY PLAN
-------------------------------------------------------------------------------------------------------------------
Bitmap Heap Scan on test_ex2  (cost=4.43..8.45 rows=1 width=37) (actual time=0.018..0.018 rows=1 loops=1)
Recheck Cond: (i = 100)
Heap Blocks: exact=1
->  Bitmap Index Scan on test_ex2_i_idx  (cost=0.00..4.43 rows=1 width=0) (actual time=0.014..0.014 rows=1 loops=1)
Index Cond: (i = 100)
Planning Time: 0.072 ms
Execution Time: 0.040 ms

what do you notice ?<br/>
Can you compare costs between the index scan and bitmap index scan ?

Now, force the optimizer to use another way to scan data :

In [229]:
SET enable_bitmapscan=off;

Rerun the query :

In [231]:
EXPLAIN (analyze)SELECT * FROM test WHERE  i < 5500;

5 row(s) returned.


QUERY PLAN
-------------------------------------------------------------------------------------------------------------
Seq Scan on test  (cost=0.00..20834.00 rows=701964 width=37) (actual time=0.031..112.405 rows=700000 loops=1)
Filter: (i < 5500)
Rows Removed by Filter: 300000
Planning Time: 0.086 ms
Execution Time: 158.067 ms

Compare all costs and verify if the first choice of the optimizer is justified, is it not?

Before moving to the next exercice, re-enable all scan methods :

In [232]:
SET enable_bitmapscan=on;
SET enable_indexscan=on; 
SET enable_seqscan=on;

## <span style="color:blue">EX - 5</span>

In this exercice, we will cover multi-column indexes.

In [233]:
DROP TABLE IF EXISTS testmultiindex;
CREATE TABLE testmultiindex (i integer not null, j integer not null, t text);
INSERT INTO testmultiindex SELECT i,j ,md5(i::text) FROM generate_series(1, 1000) i,generate_series(1, 1000) j;
CREATE INDEX ON testmultiindex (i, j);

Your application frequently runs the following query :<br/>
SELECT * from testmultiindex where j = x;  -- where x is any integer<br/>
You notice the query becomes slow, can you provide the execution plan : <br/>


In [234]:
explain analyze SELECT * from testmultiindex where j = 34;

5 row(s) returned.


QUERY PLAN
------------------------------------------------------------------------------------------------------------------
Seq Scan on testmultiindex  (cost=0.00..21846.00 rows=5000 width=40) (actual time=0.012..94.421 rows=1000 loops=1)
Filter: (j = 34)
Rows Removed by Filter: 999000
Planning Time: 0.123 ms
Execution Time: 94.542 ms

What do you advice to improve this query ? and justify your solution ?

The cost is very high and we browse data with a sequential scan so lot of reads. To minimize the cost I suggest to drop the index on testmultiindex (i, j) and I will create a new one only on j. Don't forget, an index must be maintained, if you don't you will impact writing performances: it is recommended to drop any useless index.

In [235]:
SELECT * from pg_indexes where  tablename='testmultiindex';

1 row(s) returned.


schemaname    tablename       indexname               tablespace    indexdef
------------  --------------  ----------------------  ------------  ----------------------------------------------------------------------------
ds2           testmultiindex  testmultiindex_i_j_idx                CREATE INDEX testmultiindex_i_j_idx ON ds2.testmultiindex USING btree (i, j)

In [236]:
drop index testmultiindex_i_j_idx;
create index test_j on testmultiindex(j);

Now your application needs to run 2 new queries :<br/>
SELECT * FROM testmultiindex where i = x AND j = y; -- where x and y are any integer <br/>
SELECT * FROM testmultiindex where i = x OR j = y;<br/>
Do you think the execution plan is still optimal ?<br/>


In [237]:
explain analyze SELECT * FROM testmultiindex where i = 4 AND j = 234;

9 row(s) returned.


QUERY PLAN
-------------------------------------------------------------------------------------------------------------------
Bitmap Heap Scan on testmultiindex  (cost=19.89..2913.33 rows=1 width=41) (actual time=0.528..2.237 rows=1 loops=1)
Recheck Cond: (j = 234)
Filter: (i = 4)
Rows Removed by Filter: 999
Heap Blocks: exact=1000
->  Bitmap Index Scan on test_j  (cost=0.00..19.89 rows=995 width=0) (actual time=0.342..0.342 rows=1000 loops=1)
Index Cond: (j = 234)
Planning Time: 0.374 ms
Execution Time: 2.271 ms

In [238]:
explain analyze SELECT * FROM testmultiindex where i = 4 OR j = 234;

5 row(s) returned.


QUERY PLAN
------------------------------------------------------------------------------------------------------------------
Seq Scan on testmultiindex  (cost=0.00..24346.00 rows=1989 width=41) (actual time=0.042..82.243 rows=1999 loops=1)
Filter: ((i = 4) OR (j = 234))
Rows Removed by Filter: 998001
Planning Time: 0.053 ms
Execution Time: 82.443 ms

For both queries, the execution plan shows a high cost for scanning the i column which is without index.

What do you suggest ?

I would suggest to create an index on i in addition to j. It would not be interesting to create a multi-column index (i,j) because for the predicat  i = x OR j = y, it would not be relevant.

In [ ]:
CREATE INDEX ON test_idx_i (i);
CREATE INDEX ON test_idx_j (j);